Add basic build cancellation

pwaller opened the pull request:
Cancelling the build is a desirable function. For example, let's say one wanted
to hook up inotify to trigger a build/run cycle when code was edited. At the
moment if there were multiple changes in quick succession, you would end up
with many in-flight builds contending with each other, and no way to cancel
them other than restarting the docker daemon.

This PR has been optimised for being the smallest useful change. It does not
represent the full changes needed to cancel in all circumstances yet. That
will come later, first I hope to disseminate some understanding of how to
achieve this.

Currently it only allows cancellation during the command running phase. If the
command is a `RUN` statement, the container is killed. Other statements are
not currently interrupted but instead the build is cancelled just before
the next statement.

Cancellation can be used by non-docker clients by cancelling the http request.
There is an example of how to do that here:

https://github.com/fsouza/go-dockerclient/pull/183

---

Add the capability to cancel the build by disconnecting the client.

This adds a `cancelled` channel which is used to signal that a build
should halt. The build is halted by sending a Kill signal and noticing
that the cancellation channel is closed.

This first pass implementation does not allow cancellation during a
pull, but that will come in a subsequent PR.

See also:
https://github.com/docker/docker/issues/3654
https://github.com/docker/docker/issues/6928

Next on my TODO list will be to implement the same thing in a similar style
for `docker pull`, if this PR is accepted. That change will automatically
make the pull phase of builds cancellable, too.

Docker-DCO-1.1-Signed-off-by: Peter Waller p@pwaller.net (github: pwaller)

crosbymichael commented:
I think it's an awesome idea and simple implementation.  Lets do it.

I'm thinking of the best way to handle the racy job close, I'm leaning towards a sync.Once like you said.

pwaller commented:
@SvenDowideit does this require an API version number bump?

pwaller commented:
I've added docs to the cli reference.

Does anyone have any opinion on how to implement a test for this? It's not clear that it's possible to do it reliably without introducing some way of listing on-going builds.

crosbymichael commented:
Maybe for a test we could do a `RUN sleep 5` and cancel during that sleep.  It may work but could be racy.  It's a little hard to have a test.

pwaller commented:
Hm. One possibility is to have a process which blocks indefinitely. Then it would always be obvious if it halted due to the cancel or not. The downside is that if the test is broken (because the cancellation feature is broken), the docker daemon will have a stuck process. The test itself could have a long timeout to ensure that it doesn't cause builders to be stuck indefinitely.

crosbymichael commented:
The tests have a timeout so I think that would work.  When we cancel the build we send a quit any current processes and that will make it unblock.  Sounds much better to me.

SvenDowideit commented:
@pwaller the API doc for next release is the 1.17 one, so you should be good to add there

pwaller commented:
@SvenDowideit I'm done with the documentation as far as I'm concerned, PTAL

Everyone: I'm going to write a test for this, but I'm a little busy and it might take up to ~1 week.

thaJeztah commented:
@pwaller drone complains that you need do gofmt `builder/job.go`

pwaller commented:
Okay, I just added 9fb1e21, which is a first go at testing this. I would appreciate anyone cross-checking my race analysis. I would love to know if `docker events` is well behaved [with respect to the start epoch](https://github.com/pwaller/docker/commit/9fb1e211e1da5417836814af7f668fe22d708279#diff-64d1c3a97fd8b1abf170680883e81aaaR1670).

It's tricky because really the test needs to know when the container has started and when it has exited. I've used the output of `docker events` for this purpose which may be iffy. It appears to work reliably in the number of times I've tried it.

I've also tested that running the test on master fails after the default `TIMEOUT`. That's quite long, so I'm tempted to insert our own `<-time.After()` in a select with `<-eventDie` since we know it shouldn't take longer than, say, 30s.

@SvenDowideit PTAL at the documentation :book: 
@LK4D4 PTAL especially at the test, whilst wearing your concurrency wizard hat :8ball: 

Comments welcome from all!

pwaller commented:
Hmm. I didn't receive notification of the failed test. The current situation is that the test works fine when it is run alone, but when run with other tests it fails:

Works:

```diff
+time make TESTFLAGS='-v -run ^TestBuild(Can)' test-integration-cli
-time make TESTFLAGS='-v -run ^TestBuild(Can|For)' test-integration-cli
```

So it seems I still have more work to do.

Can anyone tell me how to make the daemon emit verbose debugging output during the tests?

pwaller commented:
Build passes. It was failing because it was seeing the `start` from the previous test. I've "fixed" it by making the regex more specific. I fear though that in the future someone else could come along and add another test which could break this one. So I could use some input from testing gurus, if anyone knows who to ping.

tiborvass commented:
I'd like to wait for after 1.5 to merge this, since this is just a first step.

jfrazelle commented:
maybe we should add to "next" milestone or is that confusing?

On Wed, Jan 14, 2015 at 12:54 PM, Tibor Vass <notifications@github.com>
wrote:

> I'd like to wait for after 1.5 to merge this, since this is just a first
> step.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774#issuecomment-69989609>.
>

SvenDowideit commented:
@jfrazelle its probably better (less ambiguous) to make a `1.6` milestone or even a `consider for 1.6`

jfrazelle commented:
true trueeee

On Wed, Jan 14, 2015 at 5:46 PM, Sven Dowideit <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> its probably better (less
> ambiguous) to make a 1.6 milestone or even a consider for 1.6
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774#issuecomment-70027401>.
>

pwaller commented:
Rebased to account for new documentation and test.

What is the timeline for 1.5/1.6? When is a good time to ping the thread alive again? I assume after 1.5 is released would be a good time to start thinking seriously about this again.

MrMMorris commented:
+1 being able to Ctrl-C stuff would be lovely

pwaller commented:
Ping anyone @crosbymichael (please see my above post).

crosbymichael commented:
@pwaller I'll add this to the 1.6 milestone, thanks for working on this.

SvenDowideit commented:
Docs LGTM - @fredlf @jamtur01 

I wonder if its really desirable for the build to stop when my internet connection times out on my mobile phone connection to the remote swarm - but perhaps that's a corner case.

pwaller commented:
@SvenDowideit isn't that the sort of thing you want to solve with a tmux or some other solution? Because you probably want the output to still go somewhere useful in that case, in case of build failure, for example.

I strongly think that disconnecting from the docker server should usually be interpreted as a cancellation signal.

thaJeztah commented:
> I strongly think that disconnecting from the docker server should usually be interpreted as a cancellation signal.

Hm, not sure. Dropping connection is (in most cases?) not an explicit action by the user.

Is it *possible* to distinguish an explicit `CTRL+C` from a dropped connection? If it is, I'd be +1 on only canceling on an explicit `CTRL+C`.

pwaller commented:
@thaJeztah The way I look at the problem is this: what should happen if the docker client process is killed or un-cleanly terminated?

At the moment, a build you can't see or synchronise with continues in the background. It may more may not complete at some point in the future and make a new tag.

Most users at the moment, if they're closing their docker client during a build, it's presumably because the build needs tweaking. If they cancel, tweak and resubmit, then they've now got two builds (one invisible) racing each other.

The only way I'm aware of to respond to a remote processes being killed or un-cleanly shutting down is by responding to the socket close. If people strongly want the ability to detach from their build process, this should be provided by some mechanism which has a reliable connection to the target API (i.e, it should at least on the same machine, e.g, it could be a tmux, or maybe the mechanism could sit separately from `docker build` in the docker server).

pwaller commented:
(But I'm not sure I see a strong argument for "oh, if my network connection dies I want the build to blunder on in the background". I use a mobile connection all the time and I don't think this behaviour is desirable. If I want to cope with the connection dropping to a remote docker server, I would use a tmux).

thaJeztah commented:
Yes, you are right. For some reason I had my wires crossed and had docker *pull* cancelation in mind (issue https://github.com/docker/docker/issues/6928). For pull, I can see it useful to have docker continue if my connection is lost. Build should indeed be cancelled.

pwaller commented:
@thaJeztah Well, I would argue the same applies for pulls, really. In my ideal world, pulls would be reference counted and when the last docker client requesting that pull disconnects the pull should be cancelled. Otherwise again, it might consume expensive resources (e.g, mobile bandwidth), if I set something going and then can't cancel it.

thaJeztah commented:
@pwaller it really depends on the situation I guess. However, I think #6928 is a more appropriate location for brainstorming on that; don't want to mess up this thread with non (directly) related discussions.

pwaller commented:
I've rebased against master and updated the location of the "What's new".

pwaller commented:
Is there any way to tell if these failures are related to my implementation?

pwaller commented:
I should add that I can't reproduce the failure locally.

duglin commented:
want to rebase and squash your commits?

pwaller commented:
Is it procedure to squash commits?

pwaller commented:
Rebased. If you see failed tests please ping the thread.

duglin commented:
Unless there's a reason to keep commit separate (like you want to isolate features for easier review), yea its pretty common to just squash 'em into one.

pwaller commented:
I've reduced it into four commits to limit the "fix foo" line-noise. But unless anyone suggests otherwise, I think leaving these commits broken up like this may be helpful for future bisection.

estesp commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1857,6 +1859,131 @@ func TestBuildForceRm(t *testing.T) {
+	defer deleteImages(name)
+
+	// (Note: one year, will never finish)
+	ctx, err := fakeContext("FROM busybox\nRUN sleep 31536000", nil)
```
that's a pretty long sleep to handle when this test doesn't work.. lots of hung Jenkins nodes right now..

ncdc replied:
Is that 36.5 days????

pwaller replied:
It's supposed to be forever. The whole point is that cancellation must work and kill it.

estesp replied:
off by 10; one year sleep! :)

pwaller replied:
Are the nodes hung because of this? I just see this:

![image](https://cloud.githubusercontent.com/assets/438648/6509128/1d91a048-c355-11e4-8284-64e50ad0ef0d.png)

pwaller replied:
Sorry, I totally mis-interpreted what I was looking at, indeed, it looks like the test is catching the fact that it my cancellation implementation isn't working.

ncdc replied:
Yeah, it's tying up the jenkins jobs

estesp replied:
hung is my slang for "lots of other PRs would like to go through CI, but most nodes are trying to run this cancellation PR".. at some point they will get killed, but this is not a nice test for an active CI infrastructure

ncdc replied:
@estesp heh, I missed adding that last 0 when using my calculator

pwaller replied:
@estesp well, it's the job of the timeout to catch if things aren't being torn down. That was discussed in this very PR here: https://github.com/docker/docker/pull/9774#issuecomment-68990842 

I apologise if this is blocking the build infrastructure, but this is catching a bug as intended.

estesp replied:
Yup, the Jenkins timeout will fire (and has for the earliest runs as of a few minutes ago)..but that means ~ 50 minutes per run if this fails to work.  Maybe you could test locally and figure out why your own 10s timeout isn't working..thanks!

pwaller replied:
Yes @estesp, I will endeavour to do so. I apologise for the noise. I was not aware that this could cause problems for anyone.

tiborvass replied:
@pwaller Can we reduce the time?

duglin replied:
Was this ever resolved? I still see the big sleep?

duglin replied:
Why can't it just be something small like 20 seconds?

jfrazelle replied:
What's the problem if it cancels anyways

pwaller replied:
The intent here really is to block forever. If one wants a timeout it would be better placed elsewhere, I think.

pwaller replied:
I need persuading that this needs changing. If there was an obvious command to substitute which would block forever (and needs minimal setup), I would use that.

pwaller replied:
That it blocked the build is unfortunate, but it was actually detecting a bug as intended.

pwaller commented:
Heads up, I've tested it locally and it passes. I apologise in advance if it freezes the builders, but it shouldn't. It froze last time due to a semantic change from the rebase. Apologies again for that!

pwaller commented:
Rebased for merge conflict in `docker_remote_api_v1.17.md`.

fredlf commented on docs/sources/reference/api/docker_remote_api_v1.17.md:
```diff
@@ -1068,6 +1068,10 @@ The archive may include any number of other files,
 
+The build can be canceled by closing the HTTP request. This also means that
+if a client disconnects because it quit or was killed, the build will be
+canceled.
```
Let's simplify/clarify and put the new information where the reader needs it: The build will also be canceled if the client drops the connection by quitting or being killed.

pwaller replied:
Fixed.

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -572,6 +572,12 @@ in cases where the same set of files are used for multiple builds. The path
+This happens if you interrupt the docker client with `ctrl-c` or if the docker
+client is killed for any reason. (Note: currently only the "run" phase of the
+build can be canceled until pull cancelation is implemented. See
+[Pull 9774](https://github.com/docker/docker/pull/9774) and followups issues.)
```
I'm leery of putting future plans into documentation. It's a real maintenance headache. How will we know to remove this note when/if #9774 is completed? We should only describe what is true now: "(Note: only the "run" phase of a build can be canceled.)"

pwaller replied:
Fixed.

pwaller commented:
I've rebased and squashed to a single commit.

Locally, master currently appears to take a long time for the tests to complete (before applying my patch). Is there a known problem?

It's in:

```
go test -test.timeout=30m github.com/docker/docker/daemon/graphdriver/devmapper
```

Please ping me if the test fails.

jfrazelle commented:
We have moved this to code review.

jfrazelle commented:
@pwaller looks like the integration-cli tests cannot be compiled due to a `syscall` see jenkins. Can you update?

fredlf commented on docs/sources/reference/api/docker_remote_api_v1.17.md:
```diff
@@ -1068,6 +1068,9 @@ The archive may include any number of other files,
 which will be accessible in the build context (See the [*ADD build
 command*](/reference/builder/#dockerbuilder)).
 
+he build will also be canceled if the client drops the connection by quitting
```
Missing "T"

fredlf commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -572,6 +572,11 @@ in cases where the same set of files are used for multiple builds. The path
 must be to a file within the build context. If a relative path is specified
 then it must to be relative to the current directory.
 
+If the docker client loses connection to the daemon, the build is canceled.
```
Capitalize Docker, please (sorry I missed this earlier, my bad).

pwaller replied:
Done, but signing off for the night now :zzz:, will return in ~10 hours.

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -572,6 +572,11 @@ in cases where the same set of files are used for multiple builds. The path
 
+If the docker client loses connection to the daemon, the build is canceled.
+This happens if you interrupt the docker client with `ctrl-c` or if the docker
+client is killed for any reason. (Note: currently only the "run" phase of the
```
If this is a note please format as one.

pwaller replied:
Done.

pwaller committed:
Implement build cancellation

Add the capability to cancel the build by disconnecting the client.

This adds a `cancelled` channel which is used to signal that a build
should halt. The build is halted by sending a Kill signal and noticing
that the cancellation channel is closed.

This first pass implementation does not allow cancellation during a
pull, but that will come in a subsequent PR.

* Add documentation of cancellation to cli and API

* Protect job cancellation with sync.Once

* Add TestBuildCancelationKillsSleep

* Add test case for build cancellation of RUN statements.

Signed-off-by: Peter Waller <p@pwaller.net>

pwaller commented:
Updated @jfrazelle.

jfrazelle commented:
cool I like how simple the implementation is, will try again tomorrow but so far seems good to me :) ping @crosbymichael too

pwaller commented:
I've added a trivial commit on the end so as not to nullify the review so far. I'll squash the commit when I next rebase or when it's ready for merging.

pwaller commented:
Tests pass locally for me, is anyone in a position to understand the windows failure? I don't use windows. @crosbymichael @jfrazelle

jfrazelle commented:
windows seems unrelated but can you rebase and we can try to get this in today for 1.6.0 (today is freeze) thanks so much @pwaller

pwaller commented:
On it, @jfrazelle.

pwaller commented:
Rebased, tested locally works. Please ping me if you see failed tests and I will respond as fast as I can, though I will have having food. I'll also be leaving ~4-5 hours after this message.

jfrazelle commented:
Awesome thanks so much, testing now!

jfrazelle commented:
LGTM, note the API docs should be in 1.18 but we can update that after the fact

pwaller commented:
Uh, are they not in [v1.18](https://github.com/pwaller/docker/blob/cancellation/docs/sources/reference/api/docker_remote_api.md#v118) already? Or were you meaning something else?

jfrazelle commented on docs/sources/reference/api/docker_remote_api_v1.17.md:
```diff
@@ -1068,6 +1068,9 @@ The archive may include any number of other files,
 which will be accessible in the build context (See the [*ADD build
 command*](/reference/builder/#dockerbuilder)).
 
+The build will also be canceled if the client drops the connection by quitting
```
this is in 1.17, but dont worry if you are busy its not a blocker

pwaller replied:
Ah, right, I'm happy to move it.

jfrazelle replied:
thanks so much :)

On Fri, Mar 20, 2015 at 12:01 PM, Peter Waller <notifications@github.com>
wrote:

> In docs/sources/reference/api/docker_remote_api_v1.17.md
> <https://github.com/docker/docker/pull/9774#discussion_r26866975>:
>
> > @@ -1068,6 +1068,9 @@ The archive may include any number of other files,
> >  which will be accessible in the build context (See the [*ADD build
> >  command*](/reference/builder/#dockerbuilder)).
> >
> > +The build will also be canceled if the client drops the connection by quitting
>
> Ah, right, I'm happy to move it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774/files#r26866975>.
>

pwaller replied:
Fixed. Good spot :)

pwaller commented:
@jfrazelle janky build passed, FWIW.

jfrazelle commented:
yay ping @crosbymichael  @LK4D4

jfrazelle commented:
hmmm this one is failing on windows "--- FAIL: TestBuildCancelationKillsSleep (0.55s)
	docker_cli_build_test.go:2049: wait failed during build run: *exec.ExitError exit status 1
	docker_cli_build_test.go:2019: docker events had bad exit status: exit status 1"

jfrazelle commented:
ah its an events think I would maybe just skip it on windows for now

jfrazelle commented:
I would use a `testRequires(t, SameHostDaemon)` for now and we can make a follow up patch to make it work on windows :)

tiborvass commented:
@pwaller @jfrazelle if that's the case, then let's add a TODO above it :)

jfrazelle commented:
yes for sure you can even assign to me

pwaller commented:
On it.

pwaller commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1924,6 +1927,134 @@ func TestBuildForceRm(t *testing.T) {
+// * When docker events sees container start, close the "docker build" command
+// * Wait for docker events to emit a dying event.
+func TestBuildCancelationKillsSleep(t *testing.T) {
+	// TODO(jfrazelle): Make this work on Windows.
```
Hope this is okay, @jfrazelle. :)

Hopefully we'll see the tests all go green.

pwaller commented:
@jfrazelle it's green :)

jfrazelle commented:
yayyyy!

On Fri, Mar 20, 2015 at 2:44 PM, Peter Waller <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> it's green :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774#issuecomment-84160619>.
>

tiborvass commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1924,6 +1927,134 @@ func TestBuildForceRm(t *testing.T) {
+	eventDie := make(chan struct{})
+
+	// Start one second ago, to avoid rounding problems
+	startEpoch := time.Now().Add(-1 * time.Second)
```
@pwaller let's just use `time.Now().Sub(time.Second)` then :)

pwaller replied:
Uh, that's not possible, [note the return types of `time.Add` and `time.Sub`](http://golang.org/pkg/time/), they are `Time` and `Duration` respectively. I need a `Time` here. If you know of another way to achieve this, by all means :)

tiborvass replied:
My bad!

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1924,6 +1927,134 @@ func TestBuildForceRm(t *testing.T) {
 	logDone("build - ensure --force-rm doesn't leave containers behind")
 }
 
+func IsKilled(err error) bool {
```
can you move this to utils.go so it can be more easily seen by others can then potentially reused?

pwaller replied:
Done (in incoming push)

duglin commented on engine/job.go:
```diff
@@ -248,3 +255,15 @@ func (job *Job) StatusCode() int {
+
+// When called, causes the Job.WaitCancelled channel to unblock.
+func (job *Job) Cancel() {
+	job.cancelOnce.Do(func() {
```
Can we get a `log.Infoln("Job canceled due to connection lost")`  type of thing here?
I was looking at the logs while I tested this and noticed that at the INFO level we print stuff about the job starting, what job it is, and it ending.  But we don't say that the engine killed it unless we turn on Debug.  For those who don't run with Debug but just INFO,I think we should be consistent and add another log line.  Then all of this info is visible at the same log level and can help figure out what actually happened.  W/o this it looks like the RUN cmd that I was on died on its own, which isn't true - the daemon killed it.

pwaller replied:
Yes. However, I think this is the wrong place to put it, because while yes, the only use of `Cancel` is if a client disconnects right now is when the http connection is lost, in the future there may be other reasons to cancel. So I've put the message at the point of connection loss (i.e, the caller of `Cancel`.)

duglin commented:
overall it looks good, just a few comments/questions.

pwaller commented:
@tiborvass  @duglin responded to comments, PTAL. I'm ~relatively unlikely~ to respond today, may respond over weekend.

duglin commented:
@pwaller there's still this one: https://github.com/docker/docker/pull/9774#discussion_r26883677

estesp commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1924,6 +1926,121 @@ func TestBuildForceRm(t *testing.T) {
+
+	select {
+	case <-time.After(10 * time.Second):
+		t.Fatal("build did not cancel in a timely fashion")
```
This comment confused me last time I reviewed this code--doesn't this Fatal actually mean "build did not **start** in a timely fashion"?  This is watching for the start event, right?  Given the code is a bit hairy, this would be good to fix.

pwaller replied:
I'm rewording and adding more comments to this region.

estesp commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -1924,6 +1926,121 @@ func TestBuildForceRm(t *testing.T) {
+	}
+
+	<-eventDie
+	// If we don't get here in a timely fashion, it wasn't killed.
```
And I know we have all tried your patience on this point :) ..but, what is the goal of not having *any* timeout, given this comment says "a timely fashion"?  If you leave the sleep {effectively forever} in the build command, why can't you have a `select` here like the above with some agreeable consensus of what "timely fashion" means?  If it can't cancel in 10 minutes, is it really a success?  8 minutes, 20 minutes..?  Without a timeout on our specific CI setup, it is a 45 minute timeout, effectively, placed right here implicitly.  So, any reason there shouldn't be an explicit one that causes a `t.Fatal` here?

pwaller replied:
You've persuaded me :)

I apologise for not responding positively on this point initially, and I did not think the status quo was an ideal solution, but I wasn't convinced anyone was making suggestions from a thoroughly reasoned standpoint (more a "he's blocking the build and it's *this sleep*!"). So I think here is the right place to do it and we should define a bound on how long Cancellation should take. I'm choosing 30 seconds for now, because some poor person might want to run the tests in a very slow environment. Usually it should seem instant, though.

jfrazelle replied:
Thanks pwaller!! Really looking forward to getting this in :)

On Saturday, March 21, 2015, Peter Waller <notifications@github.com> wrote:

> In integration-cli/docker_cli_build_test.go
> <https://github.com/docker/docker/pull/9774#discussion_r26890802>:
>
> > +	}
> > +
> > +	select {
> > +	case <-time.After(10 * time.Second):
> > +		t.Fatal("build did not cancel in a timely fashion")
> > +	case <-eventStart:
> > +	}
> > +	// Now we know the container is running
> > +
> > +	err = buildCmd.Process.Kill()
> > +	if err != nil {
> > +		t.Fatalf("error killing build command: %s", err)
> > +	}
> > +
> > +	<-eventDie
> > +	// If we don't get here in a timely fashion, it wasn't killed.
>
> You've persuaded me :)
>
> I apologise for not responding positively on this point initially, and I
> did not think the status quo was an ideal solution, but I wasn't convinced
> anyone was making suggestions from a thoroughly reasoned standpoint (more a
> "he's blocking the build and it's *this sleep*!"). So I think here is the
> right place to do it and we should define a bound on how long Cancellation
> should take. I'm choosing 30 seconds for now, because some poor person
> might want to run the tests in a very slow environment. Usually it should
> seem instant, though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774/files#r26890802>.
>

pwaller commented:
I think I have addressed things. PTAL.

pwaller commented:
@jfrazelle there were failures but they seem unrelated, please advise. :)

estesp commented:
Thanks @pwaller for reworking the timeouts and comments--much clearer and understandable now.  I'm good with this @jfrazelle ; the `TestPullEventsVerified` has some real (known) issue, so it definitely has nothing to do with your code changes.

duglin commented on engine/http.go:
```diff
@@ -38,5 +40,20 @@ func (eng *Engine) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		go func() {
+			select {
+			case <-closeNotifier.CloseNotify():
+				log.Infof("Client disconnected, cancelling job: %v", job)
```
I might be doing something wrong but I'm not seeing this in the logs when I hit ctrl-c during a build - should I?

pwaller replied:
Good catch. This is actually a different interface which isn't usually used. I've forced to put it in the usual place too, which is here:

https://github.com/pwaller/docker/commit/d20a09e06ff1a933a49feab68f4150ac6fbec203#diff-44a527ea72ea9572e1cf71c7f1e2a0f0R1098

jfrazelle commented:
It is a registry issue they are looking into it

On Saturday, March 21, 2015, Peter Waller <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> there were failures but they
> seem unrelated, please advise. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774#issuecomment-84305540>.
>

duglin commented on engine/http.go:
```diff
@@ -38,5 +40,20 @@ func (eng *Engine) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// how to use them in net/http)
+
+	// Cancel the job if the HTTP connection is closed.
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
```
Near as I can tell this code is never used - I removed it and everything still seems to work.  Do we really need it?
The reason I even thought of this was because I thought this cancel stuff would work for all http interactions but I didn't so I was investigating.  Would it make sense to make it more generic so job.Cancel() is called for all types of jobs?

pwaller replied:
Hmm. Do we know the circumstances under which Engine.ServeHTTP is used? I put it in there because I saw that it was creating a job, so it stood to reason that it must also watch the `CloseNotifier`.

duglin replied:
Not sure but looking at the stack trace when builder.run() is called I see this:
```
    /usr/local/go/src/net/http/server.go:1130 +0xbb
    /go/src/github.com/docker/docker/builder/dispatchers.go:241 +0x7c
    /go/src/github.com/docker/docker/builder/evaluator.go:328 +0xa30
    /go/src/github.com/docker/docker/builder/evaluator.go:179 +0x4b8
    /go/src/github.com/docker/docker/builder/job.go:159 +0xdad
    /go/src/github.com/docker/docker/builder/job.go:43 +0x31
    /go/src/github.com/docker/docker/engine/job.go:90 +0x936
    /go/src/github.com/docker/docker/api/server/server.go:1104 +0x12c1
    /go/src/github.com/docker/docker/api/server/server.go:1289 +0x951
    /usr/local/go/src/net/http/server.go:1265 +0x41
    /go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98
 +0x2b9
    /usr/local/go/src/net/http/server.go:1703 +0x19a
```
So, if you want to catch this in a generic way then you may need to look at those functions.
But if all you care about is stopping the build() case then let's remove the unused code for now and at least make build() clean, then we can revisit if we can to do this for all operations later (which I think would be a good idea).

pwaller commented:
@spf13, is there some implication I need to act on that you've tagged this with kind/writing?

jfrazelle commented:
nah don't worry

On Saturday, March 21, 2015, Peter Waller <notifications@github.com> wrote:

> @spf13 <https://github.com/spf13>, is there some implication I need to
> act on that you've tagged this with kind/writing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/9774#issuecomment-84428644>.
>

jfrazelle commented:
@duglin  You think this one can get in today

jfrazelle commented:
We can always make minor adjustments after instead of having to go back and
forth :)

On Saturday, March 21, 2015, Jessica Frazelle <jess@docker.com> wrote:

> nah don't worry
>
> On Saturday, March 21, 2015, Peter Waller <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> @spf13 <https://github.com/spf13>, is there some implication I need to
>> act on that you've tagged this with kind/writing?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/9774#issuecomment-84428644>.
>>
>

pwaller commented:
@duglin as requested, I removed the code from engine/http.go.

pwaller commented:
Failure is still TestPullVerified.

duglin commented:
@pwaller thanks! LGTM

jfrazelle commented:
rebuilding, and will merge after CI passes, thanks so much for being so responsive @pwaller :D

pwaller commented:
Thanks @jfrazelle and everyone who contributed! Glad to have completed this :)
